RFC.md

```diff
@@ -0,0 +1,326 @@
+
+```python
+>>> type(w)
+<class 'numpy.float64'>
+```
+
+Reductions and similar operations in NumPy return the infamous NumPy scalars.
+We'll discuss these and other NumPy quirks and how we dealt with them in the
+sequel.
+
+As expected, this layer also allows for combining NumPy code and PyTorch code.
+
+```python
+import torch
+import numpy as np
+t1 = torch.tensor([1, 3, 5])
+t2 = torch.exp(t)
+# Now say the user has some code lying around which uses NumPy:
+def fn(x, y):
+    return np.multiply(x, y).sum()
+
+result = fn(t1, t2)
+t_results = torch.empty(5, dtype=torch.float64)
+t_results[0] = result  # store the result in a torch.Tensor
+```
+
+This code mixing NumPy and PyTorch already works, as `torch.Tensor` implements
+the `__array__` method. For it to work with the compatibility layer, we would
+need to wrap and unwrap the inputs / outputs. This could be done modifying `fn`
+as
+
+```python
+def fn(x, y):
+    x = np.asarray(x)
+    y = np.asarray(y)
+    ret = np.multiply(x, y).sum()
+    return ret.tensor.numpy()
+```
+
+Note that this wrapping / unwrapping process can be easily automated via a decorator.
+Even more, if a user wants to use PyTorch as a backend in a code that mixes
+PyTorch and NumPy, it will mostly be the case that it is because they want to
+trace through that code. In that setting, TorchDynamo will be able to
+automatically do the wrapping/unwrapping.
+
+## The `torch.numpy` module
+
+The bulk of the work went into implementing a system that allows us to
+implement NumPy operations in terms of those of PyTorch. The main design goals
+were
+
+1. Implement *most* of NumPy's API
+2. Preserve NumPy semantics as much as possible
+
+We say *most* of NumPy's API, because NumPy's API is not only massive, but also
+there are parts of it which cannot be implemented in PyTorch. For example,
+NumPy has support for arrays of strings, dates, and other `dtype`s that PyTorch
+does not consider. Negative strides are other example. We put together a list
+of things that are out of the scope of this project in the
+[following issue](https://github.com/Quansight-Labs/numpy_pytorch_interop/issues/73).
+
+For the bulk of the functions, we started by prioritizing most common
+operations. Then, when bringing tests from the NumPy test suit and running
+them, we would triage and prioritize how important was to fix each failure we
+found. Iterating this process, we ended up with a small list of differences
+between the NumPy and the PyTorch API which we sorted out by hand and finished
+implementing. That list and the prioritization discussion can be found in
+[the first few posts of this issue](https://github.com/Quansight-Labs/numpy_pytorch_interop/issues/87).
+
+The second point of preserving NumPy semantics as much as possible will be used
+in the sequel to discuss some points like the default dtypes that are used
+throughout the implementation.
+
+### Annotation-based preprocessing
```

Comment:
This part is mostly about _implementation_, while the next part (`### Dtypes`) is about _user-observable behavior_. Those two topics should probably be more explicitly separated, and switched around in the document. I'll also note that there's an RFC template that would be good to adhere to, and which means some refactoring of this doc: https://github.com/pytorch/rfcs/blob/master/RFC-0000-template.md